### PR TITLE
Fix negative-index arrays

### PR DESCRIPTION
**Pull request by juancarlospaco**
- Fix negative-indexed arrays and Null-indexed arrays, like `array[-9..-1, T]` etc.
- This is already enforced for literals, but not for slices, its basically an "accepts invalid" kind of bugs.
- Stdlib is not really prepared to work with negative-indexed arrays whatsoever (can of worms).

### CODE REVIEW

**Comment by metagn** [comment]
1. What is "float based array indexing"
2. What's wrong with '\0' as an array index, it's just char(0)
3. Yeah I guess we don't *need* to support this, but you could say the same about custom index ranges in general, and if it works why break it. #13966 is an example of someone using this (`array[int8, T]` becomes `array[-128..127, T]`), which the way you implemented this PR, doesn't seem like it will give an error.

File: compiler/semdata.nim [hunk 1]
```diff
@@ -103,7 +103,7 @@ type
     imports*: seq[ImportedModule] # scope for all imported symbols
     topLevelScope*: PScope     # scope for all top-level symbols
     p*: PProcCon               # procedure context
-    intTypeCache*: array[-5..32, PType] # cache some common integer types
+    intTypeCache*: array[27, PType] # cache some common integer types
```
Comment (metagn) [inline compiler/semdata.nim:106]: This is an actual use case for negative indexed arrays. We map values literally from the integers `-5, -4, -3, ..., 30, 31, 32` to type values.

You could either make this `0 .. 32` and lose the optimization for -5 .. -1, or you could set the length to be `32 - (-5)` (which is not 27) and then manually write `-5` for `low(intTypeCache)` and manually do `- (-5)` when doing the index access. If only we had a way to express this with the type system

**Comment by metagn** [comment]
Sidenote about "standard library is not prepared": This has nothing to do with negative indexes. Any problem negative indexed arrays have are shared with any custom index ranges for arrays that do not start with 0. There's really only one problem too, it's that we have implicit conversions to `openarray` and openarrays lose the `low(arr)` information, which causes problems when returning index values. Relevant discussion is at https://github.com/nim-lang/RFCs/issues/184

**Comment by Araq** [comment]
Before you spend more time on this: Rejected, nobody asked for it.

**Comment by juancarlospaco** [comment]
@Araq this is already enforced for single literals, but not for slice of literals,
without a warning, I just wanted to add a warning, not cut compilation, 
if we do support negative index then whats the point of `BackwardIndex` to exist?,
and some functions of stdlib like macros and `strutils` do not expect negative index whatsoever.

**Comment by metagn** [comment]
Negative indexes have nothing to do with backwards indexes. It's an offset for the beginning index of arrays, not from the end like in Python. It might have been this way very early on in the language but right now that's the only way it's used.

> this is already enforced for single literals, but not for slice of literals,

Really curious what you mean. Are you talking about negative *lengths* being illegal? Because again, not the same thing

Just to demonstrate, an example:

```nim
var a: array[-2 .. 2, string]
a[-2] = "a"
a[-1] = "b"
a[0] = "c"
a[1] = "d"
a[2] = "e"

var b: array[0 .. 4, string]
b[0] = "a"
b[1] = "b"
b[2] = "c"
b[3] = "d"
b[4] = "e"

var c: array[2 .. 6, string]
c[2] = "a"
c[3] = "b"
c[4] = "c"
c[5] = "d"
c[6] = "e"

echo a # ["a", "b", "c", "d", "e"]
echo b # ["a", "b", "c", "d", "e"]
echo c # ["a", "b", "c", "d", "e"]
```

This is not unheard of. You can do the same in Ada.

**Pull request closed without merging**